[docs]: add JavaScript in Gitpod guide. (#524)
gitpod-io/website: src/docs/languages/javascript.md

# JavaScript in Gitpod

Gitpod comes with great built-in support for JavaScript and tools like Node, npm, and Yarn pre-installed. Still, depending on your project, you might want to further optimize the experience.

## Examples

Here are a few JavaScript example projects that are automated with Gitpod:

<div class="table-container">

Repository | Description | Try it
---|---|---
[MobX](https://github.com/mobxjs/mobx) | Simple, scalable state management. | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/mobxjs/mobx)
[Mozilla pdf.js](https://github.com/mozilla/pdf.js) | PDF.js is a Portable Document Format (PDF) viewer that is built with HTML5. | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/mozilla/pdf.js)
[Tesseract.js](https://github.com/naptha/tesseract.js) | Pure Javascript OCR for more than 100 Languages. | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/naptha/tesseract.js)
</div>

## Start tasks

> Almost all of JavaScript projects these days use some sort of build tool for things like bundling, linting, code-splitting and so on and they also use a package manager, typically either npm or Yarn for managing dependencies. 

Reviewer: I feel it would be "safer" to say "many" instead of "almost all" (e.g. if your JavaScript project doesn't use a build tool, that's okay too, we don't want to give you the impression that "you're doing it wrong").

```suggestion
Many JavaScript projects these days use some sort of build tool for things like bundling, linting, code-splitting and so on and they also use a package manager, typically either npm or Yarn for managing dependencies. 
```

> Almost the entirety of JavaScript projects these days use some sort of build tool for things like bundling, linting, code-splitting and so on and they also use a package manager, either it is Npm or Yarn for managing dependencies. 

Reviewer 2: 'Npm or Yarn' --> 'npm or Yarn'
'Almost the entirety of' --> 'Almost all'
'either it is' --> 'typically either' (flows better and also there are other package managers such as bower for the web yarn and npm are just the most common)

You can automate the process of installing dependencies and starting any tasks like `build`, `lint`, `test` and so on at the workspace startup, for doing so please create a [.gitpod.yml](/docs/config-gitpod-file/) file in the root of your project and add the tasks you want to be automated. An example might look like this:

```yaml
tasks:
    - init: npm install && npm run build
      command: npm run dev
```

<span aria-hidden="true">👆</span> In the above example, we are telling Gitpod to run what is in the `init` phase at the time of workspace initialization and then afterwards run whatever is in the `command` phase. 

You can read more about start tasks [here](/docs/config-start-tasks/). 

## Node Versions

> Gitpod comes with node version 12.16.3 pre-installed but let's say your project uses a different version of node(let say 8 for example), well the good news is that Gitpod also comes with nvm(a tool used to manage multiple active Node versions) installed. To install and configure the desired version of node create a [.gitpod.Dockerfile](/docs/config-docker/) and add the following to it:

Reviewer: Maybe it's safer to be more vague about the Node.js version that's installed, because it tends to change frequently (e.g. when Node.js releases updates) and we'll likely forget to update this page every time.

```suggestion
Gitpod comes with the latest stable Node.js version pre-installed but let's say your project uses a different version of `node` (say 8 for example), well the good news is that Gitpod also comes with `nvm` (a tool used to manage multiple active Node.js versions) installed. To install and configure the desired version of node create a [.gitpod.Dockerfile](/docs/config-docker/) and add the following to it:
```

```dockerfile

> FROM gitpod/workspace-full-vnc:latest

Reviewer: I don't think the default example should have the VNC remote desktop (typically not very useful for JS projects, unless there is some sort of graphical app involved).

```suggestion
FROM gitpod/workspace-full:latest
```

RUN bash -c ". .nvm/nvm.sh \
    && nvm install 8 \
    && nvm use 8 \
    && nvm alias default 8"

RUN echo "nvm use default &>/dev/null" >> ~/.bashrc.d/51-nvm-fix
```

> and then in your [.gitpod.yml](/docs/config-gitpod-file/) reference [.gitpod.Dockerfile](/docs/config-docker/) as shown below:

Reviewer: ```suggestion
and then in your [.gitpod.yml](/docs/config-gitpod-file/) reference your [.gitpod.Dockerfile](/docs/config-docker/) as shown below:
```

```yaml
image: 

>     file: .gitpod.Dockerfile

Reviewer: I think it's only 2 spaces in YAML:

```suggestion
  file: .gitpod.Dockerfile
```

```

Then, after committing your changes, start a new workspace, the version for that workspace will be what you’ve specified in your [.gitpod.Dockerfile](/docs/config-docker/).

## Using Eslint for linting

> If your project's `package.json` does not mention Eslint as a dependency then you have to install it first. For installing it add the following to the end of the `- init` phase of your [.gitpod.yml](/docs/config-gitpod-file/) as shown:

Reviewer 2: why is it the '`- init` phase'
 instead of the 
'`init` phase' AFAIK the '- ' is just syntax

```yaml
tasks:
    - init: npm install && npm run build && npm install -g eslint
```

> and then also add the following in .gitpod.yml as well:

Reviewer 2: 'also and 'as well' mean the same thing in this sentence holds the same meaning in this context,

> and then also add the following in .gitpod.yml as well:

Reviewer 2: 'also and 'as well' mean the same thing in this sentence holds the same meaning in this context,

```yaml
vscode:
  extensions:
    - dbaeumer.vscode-eslint@2.0.15:/v3eRFwBI38JLZJv5ExY5g==

> ```

Reviewer: Maybe it's better to tell people to search for "eslint" in the IDE's extension panel, and to install it for their project, instead of asking them to add this "magic value" into their `.gitpod.yml`?

Author: yeah, this definitely makes more sense. Thanks
